doc/drop_counters/drop_counters_HLD.md

```diff
@@ -0,0 +1,383 @@
+-------   --------  -----  -------------  -------  -----------
+DEBUG_3   TX_LEGIT  LEGIT  SWITCH_EGRESS  L2_ANY   Legitimate switch-level TX pipeline drops
+                                          L3_ANY
+
+$ config drops add --counter="DEBUG_3" --reasons=["A_CUSTOM_REASON", "ANOTHER_CUSTOM_REASON"]
+Configuring DEBUG_3...
+
+Counter   Alias     Group  Type           Reasons                Description
+-------   --------  -----  -------------  -------                -----------
+DEBUG_3   TX_LEGIT  LEGIT  SWITCH_EGRESS  L2_ANY                 Legitimate switch-level TX pipeline drops
+                                          L3_ANY
+                                          A_CUSTOM_REASON
+                                          ANOTHER_CUSTOM_REASON
+
+$ config drops remove --counter="DEBUG_3" --reasons=["A_CUSTOM_REASON"]
+Configuring DEBUG_3...
+
+Counter   Alias     Group  Type           Reasons                Description
+-------   --------  -----  -------------  -------                -----------
+DEBUG_3   TX_LEGIT  LEGIT  SWITCH_EGRESS  L2_ANY                 Legitimate switch-level TX pipeline drops
+                                          L3_ANY
+                                          ANOTHER_CUSTOM_REASON
+
+$ config drops delete --counter="DEBUG_3"
+```
+
+## 3.2 Config DB
+Two new tables will be added to Config DB:
+* DEBUG_COUNTER to store general debug counter metadata
+* DEBUG_COUNTER_DROP_REASON to store drop reasons for debug counters that have been configured to track packet drops
+
+### 3.2.1 DEBUG_COUNTER Table
+Example:
+```
+{
+    "DEBUG_COUNTER": {
+        "DEBUG_0": {
+            "alias": "PORT_RX_LEGIT",
+            "type": "PORT_INGRESS_DROPS",
+            "desc": "Legitimate port-level RX pipeline drops",
+            "group": "LEGIT"
+        },
+        "DEBUG_1": {
+            "alias": "PORT_TX_LEGIT",
+            "type": "PORT_EGRESS_DROPS",
+            "desc": "Legitimate port-level TX pipeline drops"
+            "group": "LEGIT"
+        },
+        "DEBUG_2": {
+            "alias": "SWITCH_RX_LEGIT",
+            "type": "SWITCH_INGRESS_DROPS",
+            "desc": "Legitimate switch-level RX pipeline drops"
+            "group": "LEGIT"
+        }
+    }
+}
+```
+
+### 3.2.2 DEBUG_COUNTER_DROP_REASON Table
+Example:
+```
+{
+    "DEBUG_COUNTER_DROP_REASON": {
+        "DEBUG_0|SMAC_EQUALS_DMAC": {},
+        "DEBUG_0|INGRESS_VLAN_FILTER": {},
+        "DEBUG_1|EGRESS_VLAN_FILTER": {},
+        "DEBUG_2|TTL": {},
+    }
+}
+```
+
+## 3.3 State DB
+State DB will store information about:
+* What types of drop counters are available on this device
+* How many drop counters are available on this device
+* What drop reasons are supported by this device
+
+### 3.3.1 DEBUG_COUNTER_CAPABILITIES Table
+Example:
+```
+{
+    "DEBUG_COUNTER_CAPABILITIES": {
+        "SWITCH_INGRESS_DROPS": {
+            "total": 3,
+            "used":  1,
+            "reasons": [L2_ANY, L3_ANY, SMAC_EQUALS_DMAC]
```

Comment:
What is the scope of L2_ANY drop reason? is this specific hardware platform? what are the SONiC platforms supported this?

Reply:
L2_ANY refers to any drop reason coming from layer 2, so for example if a frame has the same source mac and destination mac, if a frame is larger than the MTU, etc. Any packet dropped due to problems at layer 2 will be counted under L2_ANY.

L2_ANY is defined in the SAI so in theory it should not be platform specific. That being said, I don't think any of the platforms support this drop reason yet except for maybe Mellanox (let me get back to you on that!)